service/olm/olm_operator.go

```diff
@@ -0,0 +1,568 @@
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"path"
+	"strings"
+	"time"
+
+	operatorsv1 "github.com/operator-framework/api/pkg/operators/v1"
+	"github.com/operator-framework/api/pkg/operators/v1alpha1"
+	operators "github.com/operator-framework/operator-lifecycle-manager/pkg/package-server/apis/operators"
+	controllerv1beta1 "github.com/percona-platform/dbaas-api/gen/controller"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	dbaascontroller "github.com/percona-platform/dbaas-controller"
+	"github.com/percona-platform/dbaas-controller/service/k8sclient"
+)
+
+const (
+	olmRepo              = "operator-lifecycle-manager"
+	githubAPIURLTemplate = "https://api.github.com/repos/operator-framework/%s/releases/latest"
+	baseDownloadURL      = "github.com/operator-framework/operator-lifecycle-manager/releases/download"
+	olmNamespace         = "olm"
+
+	// If version is not set, DBaaS controller will choose the latest from the repo.
+	// It doesn't work for offline installation.
+	latestOLMVersion  = "latest"
+	defaultOLMVersion = ""
+
+	// APIVersionCoreosV1 constant for some API requests.
+	APIVersionCoreosV1 = "operators.coreos.com/v1"
+)
+
+// ErrEmptyVersionTag Got an empty version tag from GitHub API.
+var ErrEmptyVersionTag error = errors.New("got an empty version tag from Github")
+
+// OperatorService holds methods to handle the OLM operator.
+type OperatorService struct {
+	kubeConfig string
+}
+
+// NewOperatorService returns new OperatorService instance.
+func NewOperatorService() *OperatorService {
+	return new(OperatorService) //nolint:exhaustruct
+}
+
+// NewOperatorServiceFromConfig returns new OperatorService instance and intializes the config.
+func NewOperatorServiceFromConfig(kubeConfig string) *OperatorService {
+	return &OperatorService{ //nolint:exhaustruct
+		kubeConfig: kubeConfig,
+	}
+}
+
+// InstallOLMOperator installs the OLM in the Kubernetes cluster.
+func (o *OperatorService) InstallOLMOperator(ctx context.Context, req *controllerv1beta1.InstallOLMOperatorRequest) (*controllerv1beta1.InstallOLMOperatorResponse, error) {
+	client, err := k8sclient.New(ctx, req.KubeAuth.Kubeconfig)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	defer client.Cleanup() //nolint:errcheck
+
+	response := new(controllerv1beta1.InstallOLMOperatorResponse)
+
+	if isInstalled(ctx, client, olmNamespace) {
+		return response, nil // No error, already installed.
+	}
+
+	var crdFile, olmFile interface{}
+
+	switch strings.ToLower(req.Version) {
+	case latestOLMVersion:
+		latestVersion, err := getLatestVersion(ctx, olmRepo)
+		if err != nil {
+			return nil, err
+		}
+		crdFile = "https://" + path.Join(baseDownloadURL, latestVersion, "crds.yaml")
+		olmFile = "https://" + path.Join(baseDownloadURL, latestVersion, "olm.yaml")
```

Comment:
yeah, default olm would come with default CatalogSource. We would need to switch it to the https://github.com/percona/dbaas-catalog/blob/percona-platform/percona-dbaas-catalog.yaml. Don't know if we plan it now or later?